File: demo-app-objc/CloudXObjCRemotePods/DemoAppLogger.m
#import "DemoAppLogger.h"
#import <CloudXCore/CloudXCore.h>

@implementation DemoAppLogEntry

- (instancetype)initWithMessage:(NSString *)message {
    self = [super init];
    if (self) {
        _message = [message copy];
        _timestamp = [NSDate date];
        
        NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
        formatter.dateFormat = @"HH:mm:ss.SSS";
        _formattedTimestamp = [formatter stringFromDate:_timestamp];
    }
    return self;
}

@end

@interface DemoAppLogger ()
@property (nonatomic, strong) NSMutableArray<DemoAppLogEntry *> *logs;
@property (nonatomic, strong) dispatch_queue_t logQueue;
@end

@implementation DemoAppLogger

+ (instancetype)sharedInstance {
    static DemoAppLogger *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[DemoAppLogger alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _logs = [[NSMutableArray alloc] init];
        _logQueue = dispatch_queue_create("com.cloudx.demo.logger", DISPATCH_QUEUE_SERIAL);
    }
    return self;
}

- (void)logMessage:(NSString *)message {
    if (!message) return;
    
    dispatch_async(self.logQueue, ^{
        DemoAppLogEntry *entry = [[DemoAppLogEntry alloc] initWithMessage:message];
        [self.logs addObject:entry];
        
        // Also log to console for Xcode debugging
        NSLog(@"📱 [DemoApp] %@", message);
        
        // Keep only the last 500 logs to prevent memory issues
        if (self.logs.count > 500) {
            [self.logs removeObjectAtIndex:0];
        }
    });
}

- (void)logAdEvent:(NSString *)eventName ad:(nullable CLXAd *)ad {
    if (!eventName) return;
    
    NSString *adDetails = [DemoAppLogger formatAdDetails:ad];
    NSString *fullMessage = [NSString stringWithFormat:@"%@%@", eventName, adDetails];
    [self logMessage:fullMessage];
}

- (void)clearLogs {
    dispatch_async(self.logQueue, ^{
        [self.logs removeAllObjects];
    });
}

- (NSArray<DemoAppLogEntry *> *)getAllLogs {
    __block NSArray<DemoAppLogEntry *> *logsCopy;
    dispatch_sync(self.logQueue, ^{
        logsCopy = [self.logs copy];
    });
    return logsCopy;
}

- (NSInteger)logCount {
    __block NSInteger count;
    dispatch_sync(self.logQueue, ^{
        count = self.logs.count;
    });
    return count;
}

+ (NSString *)formatAdDetails:(nullable CLXAd *)ad {
    if (!ad) {
        return @" - Ad: (null)";
    }
    
    NSMutableString *details = [NSMutableString stringWithString:@" - Ad Details:"];
    
    // Ad Unit ID
    if (ad.adUnitId) {
        [details appendFormat:@"\n  📍 Ad Unit ID: %@", ad.adUnitId];
    } else {
        [details appendString:@"\n  📍 Ad Unit ID: (null)"];
    }
    
    // Network Name
    if (ad.networkName) {
        [details appendFormat:@"\n  🏢 Network: %@", ad.networkName];
    } else {
        [details appendString:@"\n  🏢 Network: (null)"];
    }
    
    // Network Placement
    if (ad.networkPlacement) {
        [details appendFormat:@"\n  🔗 Network Placement: %@", ad.networkPlacement];
    } else {
        [details appendString:@"\n  🔗 Network Placement: (null)"];
    }
    
    // Revenue
    if (ad.revenue) {
        [details appendFormat:@"\n  💰 Revenue: $%.6f", [ad.revenue doubleValue]];
    } else {
        [details appendString:@"\n  💰 Revenue: (null)"];
    }
    
    return [details copy];
}

@end

File: demo-app-swift/CloudXSwiftRemotePods/DemoAppLogger.swift
import Foundation
import CloudXCore

struct DemoAppLogEntry {
    let message: String
    let timestamp: Date
    let formattedTimestamp: String
    
    init(message: String) {
        self.message = message
        self.timestamp = Date()
        
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm:ss.SSS"
        self.formattedTimestamp = formatter.string(from: timestamp)
    }
}

class DemoAppLogger {
    static let sharedInstance = DemoAppLogger()
    
    private var logs: [DemoAppLogEntry] = []
    private let logQueue = DispatchQueue(label: "com.cloudx.demo.logger", qos: .utility)
    
    private init() {}
    
    func logMessage(_ message: String?) {
        guard let message = message else { return }
        
        logQueue.async { [weak self] in
            guard let self = self else { return }
            
            let entry = DemoAppLogEntry(message: message)
            self.logs.append(entry)
            
            // Also log to console for Xcode debugging
            print("📱 [DemoApp] \(message)")
            
            // Keep only the last 500 logs to prevent memory issues
            if self.logs.count > 500 {
                self.logs.removeFirst()
            }
        }
    }
    
    func logAdEvent(_ eventName: String?, ad: CLXAd?) {
        guard let eventName = eventName else { return }
        
        let adDetails = DemoAppLogger.formatAdDetails(ad)
        let fullMessage = "\(eventName)\(adDetails)"
        logMessage(fullMessage)
    }
    
    func clearLogs() {
        logQueue.async { [weak self] in
            self?.logs.removeAll()
        }
    }
    
    func getAllLogs() -> [DemoAppLogEntry] {
        return logQueue.sync {
            return Array(logs)
        }
    }
    
    var logCount: Int {
        return logQueue.sync {
            return logs.count
        }
    }
    
    static func formatAdDetails(_ ad: CLXAd?) -> String {
        guard let ad = ad else {
            return " - Ad: (null)"
        }
        
        var details = " - Ad Details:"
        
        // Ad Unit Name
        if let adUnitName = ad.adUnitName {
            details += "\n  📍 Ad Unit: \(adUnitName)"
        } else {
            details += "\n  📍 Ad Unit: (null)"
        }
        
        // Ad Unit ID
        if let adUnitId = ad.adUnitId {
            details += "\n  🆔 Ad Unit ID: \(adUnitId)"
        } else {
            details += "\n  🆔 Ad Unit ID: (null)"
        }

        // Placement (publisher-provided)
        if let placement = ad.placement {
            details += "\n  📌 Placement: \(placement)"
        } else {
            details += "\n  📌 Placement: (null)"
        }

        // Network Name
        if let networkName = ad.networkName {
            details += "\n  🏢 Network: \(networkName)"
        } else {
            details += "\n  🏢 Network: (null)"
        }
        
        // Network Placement
        if let networkPlacement = ad.networkPlacement {
            details += "\n  🔗 Network Placement: \(networkPlacement)"
        } else {
            details += "\n  🔗 Network Placement: (null)"
        }
        
        // Revenue
        if let revenue = ad.revenue {
            details += String(format: "\n  💰 Revenue: $%.6f", revenue.doubleValue)
        } else {
            details += "\n  💰 Revenue: (null)"
        }
        
        return details
    }
}